### PR TITLE
[Bounds-Safety] Add sized_by, counted_by_or_null & sized_by_or_null

### PR DESCRIPTION
**Pull request by hnrklssn**
The attributes `sized_by`, `counted_by_or_null` and `sized_by_or_null` have been added as variants on `counted_by`, each with slightly different semantics.  `sized_by` takes a byte size parameter instead of an element count, allowing pointees with unknown size. The `counted_by_or_null` and `sized_by_or_null` variants are equivalent to their base variants, except the pointer can be null regardless of count/size value.  If the pointer is null the size is effectively 0.

rdar://125400354

### CODE REVIEW

**Comment by hnrklssn** [comment]
Blocked by pointer support for `counted_by` reverted. It's being relanded in #93121

**Comment by delcypher** [comment]
@hnrklssn #93121 has been landed. Let's hope it sticks this time 🤞

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (delcypher) [inline clang/lib/Sema/SemaDeclAttr.cpp:8652]: Rather than magic constants it might be better to use an enum that you refer to here and in `DiagnosticSemaKinds.td` so that it's more readable.

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (delcypher) [inline clang/lib/Sema/SemaDeclAttr.cpp:8712]: I would have thought we would want to prevent `sized_by` on all of the conditions above (e.g. pointee being a function type) apart from pointee being an incomplete type. However, it looks like this implementation allows pointee types for all the cases we disallow it for `counted_by`.

**Review by delcypher — CHANGES_REQUESTED** [review]
Looks pretty good. I have some minor comments.

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (hnrklssn) [reply clang/lib/Sema/SemaDeclAttr.cpp:8712]: I think that is a reasonable restriction to add in `-fbounds-safety`, but for simply saying "this is the size", is it really an issue? It's also feasible to allow the attribute, but not allow indexing into an unsized type. That would make it possible to cast to `char * sized_by(sz)` and retain the size bounds before indexing into it.

File: clang/docs/ReleaseNotes.rst
Comment (hnrklssn) [reply clang/docs/ReleaseNotes.rst:432]: Hmm yeah there's nothing preventing you from assigning null I guess, but I don't believe `__bdos` CG has any support for returning 0 when the pointer is null.

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (hnrklssn) [reply clang/lib/Sema/SemaDeclAttr.cpp:8663]: > The OrNull case probably deserves its own special diagnostic because in that case the diagnostic should explain that they cannot use the attribute on arrays and that they need to use __counted_by instead.

That is also true for `CountInBytes`. But I agree that it would be good for the diagnostic to suggest `counted_by`.

Comment (hnrklssn) [reply clang/lib/Sema/SemaDeclAttr.cpp:8663]: >The diagnostic name is a little misleading here because CountInBytes suggested __sized_by but the diagnostic name has counted_by in its name

I kept it because it's the same family of attributes. Do you have a suggestion for a name that would imply that it's not just for `counted_by`, but more specific than `bounds_attribute`?

Comment (delcypher) [reply clang/lib/Sema/SemaDeclAttr.cpp:8663]: > > The OrNull case probably deserves its own special diagnostic because in that case the diagnostic should explain that they cannot use the attribute on arrays and that they need to use __counted_by instead.
> 
> That is also true for `CountInBytes`. But I agree that it would be good for the diagnostic to suggest `counted_by`.

Hmm looking at it more closely for the `err_counted_by_attr_not_on_ptr_or_flexible_array_member` means the diagnostic output for

```
struct Test {
  int count;
  int fma[] __counted_by_or_null(count)
}
```

would look something like

```
counted_by_or_null only applies to pointers
```

i.e. mentioning flexible array members is dropped (I initially thought it wasn't). So actually I guess this is fine. Suggesting the right attribute would be an extra bonus.  If you don't want to do it now file an issue (tagged with clang:bounds-safety).

Comment (delcypher) [reply clang/lib/Sema/SemaDeclAttr.cpp:8663]: > > The diagnostic name is a little misleading here because CountInBytes suggested __sized_by but the diagnostic name has counted_by in its name
> 
> I kept it because it's the same family of attributes. Do you have a suggestion for a name that would imply that it's not just for `counted_by`, but more specific than `bounds_attribute`?

We use `CountAttributedType` to represent both `counted_by` and `sized_by`  so how about `err_count_attr_...` as the diagnostic prefix? I don't have strong opinions on exactly what the name should be.

**Review by Endilll — COMMENTED** [review]
`Sema.h` changes look good.

**Review by AaronBallman — APPROVED** [review]
There are a few minor suggestions from @delcypher that should be addressed, but overall the changes LGTM (assuming there are no precommit CI surprises after fixing the merge conflicts).

File: clang/docs/ReleaseNotes.rst
Comment (rapidsna) [reply clang/docs/ReleaseNotes.rst:432]: It would be nice to add a suggested use case for `sized_by_or_null`. e.g., an allocator that returns either a buffer of size or nullptr.

**Review by rapidsna — APPROVED** [review]

**Comment by hnrklssn** [comment]
Rebased on main after @delcypher's patch was relanded. Addressing comments now.

**Review by AaronBallman — APPROVED** [review]
Continues to LGTM, @delcypher are you happy with the changes?

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (delcypher) [reply clang/lib/Sema/SemaDeclAttr.cpp:8416]: > I think that is a reasonable restriction to add in -fbounds-safety, but

I think we should avoid bifurcating the semantics of `sized_by` based on the `-fbounds-safety` flag where possible. The more differences we have the more pain we are going to cause during adoption. Consider a project that first annotates their headers with `sized_by` but doesn't pass `-fbounds-safety`. They will get one set of semantics, then a project that consumes that header file builds with `-fbounds-safety`. That project is built with different semantics. If the semantics differ enough its possible that the header file will build fine without `-fbound-safety` but fail to build with it.

> but for simply saying "this is the size", is it really an issue?

The issue for me is that it doesn't really make sense to have pointers to "sizeless types" (not to be confused with "incomplete types") or to "function types" in the first place and so to me that implies adding `sized_by` to those pointers doesn't make sense either.

I don't want to delay this PR for too long so if you think my ask is unreasonable we can file an issue and get back to it.

Ultimately I think @rapidsna should make the call on this issue.

**Review by delcypher — CHANGES_REQUESTED** [review]
I'm requesting changes for now but we can change this based on @rapidsna 's opinion.

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (rapidsna) [reply clang/lib/Sema/SemaDeclAttr.cpp:8416]: I agree with @delcypher.

**Comment by hnrklssn** [comment]
@delcypher @rapidsna updated __sized_by(_or_null) to only allow types with unknown size if incomplete

**Comment by rapidsna** [comment]
Still LGTM

File: clang/test/Sema/attr-counted-by-or-null-last-field.c
Comment (delcypher) [inline clang/test/Sema/attr-counted-by-or-null-last-field.c:19]: Nit: This wording is a little odd. I would expect something like.

```
'counted_by_or_null' references field 'count' which isn't within the same struct as the annotated pointer
```

**Review by delcypher — APPROVED** [review]
Thanks for addressing my comment. I have some nits but I'll leave it to your discretion whether or not you fix them.

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (hnrklssn) [reply clang/lib/Sema/SemaDeclAttr.cpp:8324]: I guess it could be a static method. It can't be an instance method since it's used before the object is constructed. Since it's only used in one place I thought it'd improve readability to keep it close to the call site.

**Comment by Xazax-hun** [comment]
I really like the test coverage of this feature, but I started to wonder if we should start splitting up some sema tests into subfolders. I think the bounds-safety related tests would be a good candidate to have their own subfolder :)

**Comment by delcypher** [comment]
@Xazax-hun This is something I'd like too. Unfortunately in previous discussions we were asked to not do this.
https://github.com/llvm/llvm-project/pull/70480#discussion_r1379166020
https://github.com/llvm/llvm-project/pull/70480#discussion_r1377966526

As the number of tests grow I think it will become more justfiable to put Bounds Safety tests in their own folder.

**Pull request merged**
